File: python/src/azoth/process/reference/packed_column.py
"""``process.packed_column`` - the packed column, which is the tray column at a derived height.

Spec: ``specs/models/process/packed_column.toml``.

The Python twin of ``crates/azoth-process/src/models/packed_column.rs``. **The arithmetic is
``process.distillation_column``'s**, because the class's is: ``PackedColumn extends
DistillationColumn`` in 463 lines and its ``run`` is ``super.run(id)`` followed by
``calcPackingHydraulics()``. What this module adds is the stage count a constructor derives from
a packed height, and the refusal of the one packing parameter the class itself refuses.

**The packing group is the excluded report's**, so four of its five parameters are carried
inertly: ``packing_type``, ``structured_packing``, ``design_flood_fraction`` and
``column_diameter`` are read by ``ColumnInternalsDesigner`` after the column has converged, and
no part of the solve reads them.
"""

from __future__ import annotations

import math

from azoth.core.errors import InvalidInputError
from azoth.core.result import PackedColumnResult
from azoth.core.units import Q
from azoth.process.reference.distillation_column import (
    distillation_column as _distillation_column,
)

#: The HETP guess ``PackedColumn``'s constructors divide a packed height by.
HETP_GUESS_M = 0.5


def stage_count(packed_height_m: float) -> int:
    """The stage count ``PackedColumn.estimateStages`` derives from a packed height.

    ``ceil(packed_height / 0.5)`` floored at two. **The floor is reached rather than refused,
    and that is measured**: the capture's stage-count rows report two middle trays for ``0.2`` m,
    for ``0.0`` m and for ``-1.0`` m alike, because neither ``estimateStages`` nor
    ``setPackedHeight`` has a domain check.

    Args:
        packed_height_m: the packed bed height, in metres.

    Returns:
        The middle trays the constructor would make.
    """
    if not math.isfinite(packed_height_m):
        # Java's `(int)` cast maps a `NaN` to zero and saturates an infinity at the type's
        # maximum; the floor decides the first, and the second is an allocation either way.
        return 2
    return max(math.ceil(packed_height_m / HETP_GUESS_M), 2)


def packed_column(
    components: list[str],
    feed_n: Q,
    feed_z: list[float],
    feed_p: Q,
    feed_t: Q,
    packed_height: Q,
    feed_stage: int,
    has_reboiler: bool,
    has_condenser: bool,
    top_pressure: Q,
    bottom_pressure: Q,
    temperature_tolerance: float = 1.0e-6,
    max_iterations: int = 200,
    reboiler_temperature: Q | None = None,
    condenser_temperature: Q | None = None,
    packing_type: str | None = None,
    structured_packing: bool | None = None,
    design_flood_fraction: float | None = None,
    packing_hydraulic_capacity_factor: float | None = None,
    column_diameter: Q | None = None,
    murphree_efficiency: float | None = None,
    solver_type: str | None = None,
    top_specification_type: str | None = None,
    top_specification_target: float | None = None,
    top_specification_component: str | None = None,
    bottom_specification_type: str | None = None,
    bottom_specification_target: float | None = None,
    bottom_specification_component: str | None = None,
) -> PackedColumnResult:
    """Solve a packed column.

    Args:
        components: the feed's substances, by name.
        feed_n: the feed's molar flow.
        feed_z: the feed composition.
        feed_p: the feed's pressure, which is its own.
        feed_t: the feed's temperature.
        packed_height: the packed bed height, which the constructor divides by a ``0.5`` m HETP
            guess to fix the stage count: ``ceil(packed_height / 0.5)`` floored at two.
        feed_stage: the stage the feed enters, 0-based over the trays including the ends.
        has_reboiler: whether stage 0 is a reboiler.
        has_condenser: whether the top stage is a condenser.
        top_pressure: the pressure at the top stage.
        bottom_pressure: the pressure at stage 0.
        temperature_tolerance: the gate on the mean tray-temperature change.
        max_iterations: the iteration cap.
        reboiler_temperature: the reboiler's temperature, which pins the bottom tray.
        condenser_temperature: the condenser's temperature, which pins the top tray.
        packing_type: the packing's name. **It does not enter the solve.**
        structured_packing: whether the packing is structured. **It does not enter the solve.**
        design_flood_fraction: the fraction of flood the packing is sized to. **It does not enter
            the solve.**
        packing_hydraulic_capacity_factor: the packing's relative hydraulic capacity factor.
            **Refused where it is not positive and finite**, which is
            ``setPackingHydraulicCapacityFactor``'s own check.
        column_diameter: the column's internal diameter. **It does not enter the solve.**
        murphree_efficiency: **not ported**; omitted is the ideal stage.
        solver_type: **only ``direct_substitution`` and ``naphtali_sandholm`` are ported**.
        top_specification_type: the top product's degree of freedom.
        top_specification_target: its target value.
        top_specification_component: the component a purity or a recovery constrains.
        bottom_specification_type: the bottom product's degree of freedom.
        bottom_specification_target: its target value.
        bottom_specification_component: the component a purity or a recovery constrains.

    Returns:
        The base column's own record, which is the id's claim.

    Raises:
        InvalidInputError: for a ``packing_hydraulic_capacity_factor`` that is not positive and
            finite, and for every input ``process.distillation_column`` refuses.

    See :func:`azoth.process.reference.distillation_column.distillation_column`.
    """
    if packing_hydraulic_capacity_factor is not None and (
        not math.isfinite(packing_hydraulic_capacity_factor)
        or packing_hydraulic_capacity_factor <= 0.0
    ):
        raise InvalidInputError(
            "packing_hydraulic_capacity_factor",
            f"a packing hydraulic capacity factor of {packing_hydraulic_capacity_factor} is not "
            "positive and finite, which `PackedColumn.setPackingHydraulicCapacityFactor` refuses "
            "with an `IllegalArgumentException`",
        )
    # The other four are declarations the solve is indifferent to: the class accepts them and
    # `ColumnInternalsDesigner` reads each one after the column has converged.
    del packing_type, structured_packing, design_flood_fraction, column_diameter

    out = _distillation_column(
        components,
        feed_n,
        feed_z,
        feed_p,
        feed_t,
        stage_count(_metres(packed_height)),
        feed_stage,
        has_reboiler,
        has_condenser,
        top_pressure,
        bottom_pressure,
        temperature_tolerance,
        max_iterations,
        reboiler_temperature,
        condenser_temperature,
        murphree_efficiency,
        solver_type,
        top_specification_type,
        top_specification_target,
        top_specification_component,
        bottom_specification_type,
        bottom_specification_target,
        bottom_specification_component,
    )
    return PackedColumnResult(
        tray_temperature=out.tray_temperature,
        tray_pressure=out.tray_pressure,
        tray_gas_n=out.tray_gas_n,
        tray_liquid_n=out.tray_liquid_n,
        distillate_n=out.distillate_n,
        distillate_z=out.distillate_z,
        distillate_p=out.distillate_p,
        distillate_t=out.distillate_t,
        distillate_h=out.distillate_h,
        bottoms_n=out.bottoms_n,
        bottoms_z=out.bottoms_z,
        bottoms_p=out.bottoms_p,
        bottoms_t=out.bottoms_t,
        bottoms_h=out.bottoms_h,
        gas_side_draw_n=out.gas_side_draw_n,
        liquid_side_draw_n=out.liquid_side_draw_n,
        pumparound_n=out.pumparound_n,
        condenser_duty=out.condenser_duty,
        reboiler_duty=out.reboiler_duty,
        iterations=out.iterations,
        temperature_residual=out.temperature_residual,
        mass_residual=out.mass_residual,
        energy_residual=out.energy_residual,
        warnings=out.warnings,
    )


def _metres(packed_height: Q) -> float:
    """The packed height in metres, which is the unit the constructor's rule is stated in."""
    return float(packed_height.to("m").magnitude)

File: crates/azoth-process/src/models/packed_column.rs
//! `process.packed_column` - the packed column as a registered id.
//!
//! Spec: `specs/models/process/packed_column.toml`. **The arithmetic is
//! [`crate::models::distillation_column`]'s**, because the class's is: `PackedColumn extends
//! `DistillationColumn`, its `run` is `super.run(id)` and the packing is read by a hydraulics
//! report afterwards. What this module adds is the stage count the class's constructor derives
//! from a packed height, and the refusal of the one packing parameter the class itself refuses.

use azoth_core::units::{MolarEnergy, Power, Pressure, ThermodynamicTemperature};
use azoth_core::{AzothError, CalcResult, Result, Warning};

use crate::kernels::packed_column::stage_count;
use crate::models::distillation_column::{DistillationColumnResult, distillation_column};

/// Result of `process.packed_column`.
///
/// **The same record as [`DistillationColumnResult`], and that is the id's claim rather than a
/// convenience**: the packing parameters reach the separation only through the stage count, and
/// the three quantities a caller would look for here instead - HETP, the theoretical stages and
/// the percent flood - are `ColumnInternalsDesigner`'s report on the far side of the solve and
/// are not ported.
#[derive(Debug, Clone, PartialEq)]
pub struct PackedColumnResult {
    /// Each tray's temperature, K.
    pub tray_temperature: Vec<ThermodynamicTemperature>,
    /// Each tray's pressure, Pa.
    pub tray_pressure: Vec<Pressure>,
    /// Each tray's vapour traffic, mol/s.
    pub tray_gas_n: Vec<f64>,
    /// Each tray's liquid traffic, mol/s.
    pub tray_liquid_n: Vec<f64>,
    /// Distillate molar flow, mol/s.
    pub distillate_n: f64,
    /// Distillate composition.
    pub distillate_z: Vec<f64>,
    /// Distillate pressure.
    pub distillate_p: Pressure,
    /// Distillate temperature.
    pub distillate_t: ThermodynamicTemperature,
    /// Distillate molar enthalpy.
    pub distillate_h: MolarEnergy,
    /// Bottoms molar flow, mol/s.
    pub bottoms_n: f64,
    /// Bottoms composition.
    pub bottoms_z: Vec<f64>,
    /// Bottoms pressure.
    pub bottoms_p: Pressure,
    /// Bottoms temperature.
    pub bottoms_t: ThermodynamicTemperature,
    /// Bottoms molar enthalpy.
    pub bottoms_h: MolarEnergy,
    /// **The vapour each tray withdrew**, which is the base column's and empty for this id: the
    /// packing does not change a draw, and its fractions are not declared here.
    pub gas_side_draw_n: Vec<f64>,
    /// The liquid each tray withdrew as a liquid side draw.
    pub liquid_side_draw_n: Vec<f64>,
    /// The liquid each tray withdrew as a pumparound.
    pub pumparound_n: Vec<f64>,
    /// The condenser's duty, W.
    pub condenser_duty: Power,
    /// The reboiler's duty, W.
    pub reboiler_duty: Power,
    /// Iterations taken.
    pub iterations: u32,
    /// The mean tray-temperature change at the last iteration, K.
    pub temperature_residual: f64,
    /// The products' worst component imbalance against the feed, relative.
    pub mass_residual: f64,
    /// The enthalpy closure.
    pub energy_residual: f64,
    /// Caveats.
    pub warnings: Vec<Warning>,
}

impl From<DistillationColumnResult> for PackedColumnResult {
    fn from(out: DistillationColumnResult) -> Self {
        Self {
            tray_temperature: out.tray_temperature,
            tray_pressure: out.tray_pressure,
            tray_gas_n: out.tray_gas_n,
            tray_liquid_n: out.tray_liquid_n,
            distillate_n: out.distillate_n,
            distillate_z: out.distillate_z,
            distillate_p: out.distillate_p,
            distillate_t: out.distillate_t,
            distillate_h: out.distillate_h,
            bottoms_n: out.bottoms_n,
            bottoms_z: out.bottoms_z,
            bottoms_p: out.bottoms_p,
            bottoms_t: out.bottoms_t,
            bottoms_h: out.bottoms_h,
            gas_side_draw_n: out.gas_side_draw_n,
            liquid_side_draw_n: out.liquid_side_draw_n,
            pumparound_n: out.pumparound_n,
            condenser_duty: out.condenser_duty,
            reboiler_duty: out.reboiler_duty,
            iterations: out.iterations,
            temperature_residual: out.temperature_residual,
            mass_residual: out.mass_residual,
            energy_residual: out.energy_residual,
            warnings: out.warnings,
        }
    }
}

impl CalcResult for PackedColumnResult {
    const CALC_ID: &'static str = "process.packed_column";
    const FIELDS: &'static [&'static str] = &[
        "tray_temperature",
        "tray_pressure",
        "tray_gas_n",
        "tray_liquid_n",
        "distillate_n",
        "distillate_z",
        "distillate_p",
        "distillate_t",
        "distillate_h",
        "bottoms_n",
        "bottoms_z",
        "bottoms_p",
        "bottoms_t",
        "bottoms_h",
        "gas_side_draw_n",
        "liquid_side_draw_n",
        "pumparound_n",
        "condenser_duty",
        "reboiler_duty",
        "iterations",
        "temperature_residual",
        "mass_residual",
        "energy_residual",
        "warnings",
    ];

    fn warnings(&self) -> &[Warning] {
        &self.warnings
    }
}

/// Solve a packed column.
///
/// # Errors
/// [`AzothError::InvalidInput`] for a `packing_hydraulic_capacity_factor` that is not positive
/// and finite, which is `setPackingHydraulicCapacityFactor`'s own check and the one parameter of
/// the packing group the class refuses rather than reads; and every error
/// [`distillation_column`] raises, which this id inherits because the solve is the base's.
#[allow(clippy::too_many_arguments)] // one parameter per declared input, and there are twenty-six
pub fn packed_column(
    components: &[String],
    feed_n: f64,
    feed_z: &[f64],
    feed_p: Pressure,
    feed_t: ThermodynamicTemperature,
    packed_height: f64,
    feed_stage: usize,
    has_reboiler: bool,
    has_condenser: bool,
    top_pressure: Pressure,
    bottom_pressure: Pressure,
    reboiler_temperature: Option<ThermodynamicTemperature>,
    condenser_temperature: Option<ThermodynamicTemperature>,
    temperature_tolerance: f64,
    max_iterations: usize,
    packing_type: Option<&str>,
    structured_packing: Option<bool>,
    design_flood_fraction: Option<f64>,
    packing_hydraulic_capacity_factor: Option<f64>,
    column_diameter: Option<f64>,
    murphree_efficiency: Option<f64>,
    solver_type: Option<&str>,
    top_specification_type: Option<&str>,
    top_specification_target: Option<f64>,
    top_specification_component: Option<&str>,
    bottom_specification_type: Option<&str>,
    bottom_specification_target: Option<f64>,
    bottom_specification_component: Option<&str>,
) -> Result<PackedColumnResult> {
    // **The class's own check, on the one packing parameter it does not simply read.**
    // `setPackingHydraulicCapacityFactor` throws for a value that is not positive and finite, so
    // the class refuses a factor its own setter would reject rather than carrying it inertly as
    // it carries the other four.
    if let Some(factor) = packing_hydraulic_capacity_factor {
        if !factor.is_finite() || factor <= 0.0 {
            return Err(AzothError::invalid_input(
                "packing_hydraulic_capacity_factor",
                format!(
                    "a packing hydraulic capacity factor of {factor} is not positive and finite, \
                     which `PackedColumn.setPackingHydraulicCapacityFactor` refuses with an \
                     `IllegalArgumentException`"
                ),
            ));
        }
    }
    // **The other four are declarations the solve is indifferent to.** The class accepts them
    // and every one of them is read by `ColumnInternalsDesigner` after the column has
    // converged, so the model carries them rather than reading them - the shape
    // `process.absorption_column` gives `max_allowable_gas_load_factor`.
    let _ = (
        packing_type,
        structured_packing,
        design_flood_fraction,
        column_diameter,
    );

    let stages = stage_count(packed_height);
    let out = distillation_column(
        components,
        feed_n,
        feed_z,
        feed_p,
        feed_t,
        stages,
        feed_stage,
        has_reboiler,
        has_condenser,
        top_pressure,
        bottom_pressure,
        reboiler_temperature,
        condenser_temperature,
        temperature_tolerance,
        max_iterations,
        murphree_efficiency,
        solver_type,
        top_specification_type,
        top_specification_target,
        top_specification_component,
        bottom_specification_type,
        bottom_specification_target,
        bottom_specification_component,
        // **A packed column's stages are equilibrium stages**: `PackedColumn` inherits
        // `setReactive`, and this id does not declare the section yet - the packing is a report
        // on the far side of the solve, and the reactive section is the base column's own.
        None,
        None,
        None,
        // **A packed column's draws are the base column's, and this id does not declare them.**
        None,
        None,
        None,
    )?;

    Ok(out.into())
}
